### ISSUE TITLE
Controller doesn't recover from "haproxy-runtime-api.sock timeout reached"

### ISSUE THREAD

**Issue by joschg** [haproxytech/kubernetes-ingress#113]
We're running the HAProxy ingress controller in our cluster on 28 pods and recently we started having some issues: some HAPoxys stop updating completely (both ingresses and backend services), they just keep running with their last updated state. In the logs we found the following line

`controller-events.go:334: /var/run/haproxy-runtime-api.sock timeout reached [set server xxxxxxxxx-xxxxxxxxxxx-xxxx/XXX_XXXXX addr xxx.xxx.xxx.xxx]`

After that there's no more log entries. The issue happens sporadically and we don't know why exactly it happens. Still though, the controller is running and appears to be healthy from the outside but fails to update and reflect any changes from the cluster. Affected instances are stuck in this state and are not able to update until they are restarted.

To me this looks like a bug rather than expected behavior: The controller fails to communicate with HAProxy but doesn't retry (at least the logs suggest that). If communication still fails after a couple of retries I'd expect the controller to exit with an error.

Kubernetes version: 1.13.1
HAProxy version: 2.0.6
HAProxy Ingress Controller version: 1.2.2

**Comment by oktalz** [comment]
hi @joschg 
yes, this should not happen and you are right, if for some reason communication is not possible it would be better to just exit, since controller cannot change it.

Usually if its fails to update it through runtime api it will try to just reload HAProxy 
(because change is also saved on drive), but since you mentioned that afterwards there are not log messages we need to see what is the issue here

**Comment by Mo3m3n** [comment]
Hi @joschg 
The controller version 1.2.2 is outdatd and a number of bugs are fixed in the dev branch.
Can you please try the dev tag and tell us if you still facing this issue ?
We are planning on making a new release 1.2.5 this monday.

**Comment by joschg** [comment]
Hi @Mo3m3n 
Thanks for looking into this! We deployed 1.2.5 today and the issue is still happening:
`controller.go:391: /var/run/haproxy-runtime-api.sock timeout reached [set server xxxxxxxx/XXX_XXXXX addr 127.0.0.1 port 8080]`
Behavior is still the same as described above.

**Comment by Mo3m3n** [comment]
Thanks for the feedback.
It would help to share numbers on cpu usage of the controller pod. Usually a timeout reaching the stat socket is caused by HAProxy being too busy to answer in time, so can you share the following for the pod showing (and during) the reported behavior:
1/ `kubectl describe pod <controller-pod>`
2/ connect to the container and run: 
`echo 'show info' | socat stdio /var/run/haproxy-runtime-api.sock`
`echo 'show activity' | socat stdio /var/run/haproxy-runtime-api.sock`

**Comment by joschg** [comment]
Hi @Mo3m3n 

Here's the output. I ran it today in the morning (~09:00 UTC) on the affected pod which had the timeout yesterday at ~13:30 UTC (we let it run until now even though it still doesn't do any updates, last log entry still the line I posted above)

```
kubectl describe pod haproxy-ingress-6d54757678-zhlz6

Name:               haproxy-ingress-6d54757678-zhlz6
Namespace:          somens
Priority:           0
PriorityClassName:  <none>
Node:               k8s-worker-25.somecomp.com/xxx.xxx.xxx.xxx
Start Time:         Mon, 18 Nov 2019 14:07:59 +0100
Labels:             date=1574072811
                    pod-template-hash=6d54757678
                    run=haproxy-ingress
Annotations:        recordset: haproxy.somecomp.com
Status:             Running
IP:                 xxx.xxx.xxx.xxx
Controlled By:      ReplicaSet/haproxy-ingress-6d54757678
Containers:
  haproxy-ingress:
    Container ID:  docker://d9c68a3fffd6c8b96ed167a7dd347f5345ecf7f30729ea75152ec24c01e2e3a9
    Image:         somecomp/haproxy-kubernetes-ingress:1.2.5
    Image ID:      docker-pullable://somecomp/haproxy-kubernetes-ingress@sha256:2eb13e73a5e035b69d9277f753edc352cfa3d8a555bd0ec00c42b558cee793eb
    Ports:         80/TCP, 443/TCP, 1024/TCP
    Host Ports:    80/TCP, 443/TCP, 1024/TCP
    Args:
      --default-ssl-certificate=default/somecomp-wildcard
      --configmap=somens/haproxy-configmap
      --default-backend-service=somens/ingress-default-backend
    State:          Running
      Started:      Mon, 18 Nov 2019 14:08:02 +0100
    Ready:          True
    Restart Count:  0
    Limits:
      memory:  16Gi
    Requests:
      memory:  16Gi
    Liveness:  tcp-socket :80 delay=15s timeout=1s period=20s #success=1 #failure=3
    Environment:
      POD_NAME:       haproxy-ingress-6d54757678-zhlz6 (v1:metadata.name)
      POD_NAMESPACE:  somens (v1:metadata.namespace)
    Mounts:
      /stats from stats (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from haproxy-ingress-service-account-token-8pmrd (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             True 
  ContainersReady   True 
  PodScheduled      True 
Volumes:
  stats:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     
    SizeLimit:  <unset>
  haproxy-ingress-service-account-token-8pmrd:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  haproxy-ingress-service-account-token-8pmrd
    Optional:    false
QoS Class:       Burstable
Node-Selectors:  type=worker
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:          <none>
```

```
echo 'show info' | socat stdio /var/run/haproxy-runtime-api.sock

Name: HAProxy
Version: 2.0.8
Release_date: 2019/10/23
Nbthread: 12
Nbproc: 1
Process_num: 1
Pid: 143
Uptime: 0d 20h05m47s
Uptime_sec: 72347
Memmax_MB: 0
PoolAlloc_MB: 4636
PoolUsed_MB: 57
PoolFailed: 0
Ulimit-n: 5000000
Maxsock: 5000000
Maxconn: 2499968
Hard_maxconn: 2499968
CurrConns: 11635
CumConns: 78116765
CumReq: 314928148
MaxSslConns: 0
CurrSslConns: 5563
CumSslConns: 33798077
Maxpipes: 0
PipesUsed: 0
PipesFree: 0
ConnRate: 904
ConnRateLimit: 0
MaxConnRate: 12940
SessRate: 904
SessRateLimit: 0
MaxSessRate: 13554
SslRate: 362
SslRateLimit: 0
MaxSslRate: 1614
SslFrontendKeyRate: 332
SslFrontendMaxKeyRate: 1756
SslFrontendSessionReuse_pct: 8
SslBackendKeyRate: 0
SslBackendMaxKeyRate: 0
SslCacheLookups: 23734468
SslCacheMisses: 21136399
CompressBpsIn: 0
CompressBpsOut: 0
CompressBpsRateLim: 0
ZlibMemUsage: 0
MaxZlibMemUsage: 0
Tasks: 36283
Run_queue: 4
Idle_pct: 86
node: k8s-worker-25.gc.mobfox.com
Stopping: 0
Jobs: 11644
Unstoppable Jobs: 0
Listeners: 8
ActivePeers: 0
ConnectedPeers: 0
DroppedLogs: 0
BusyPolling: 0
FailedResolutions: 0
TotalBytesOut: 948686008567
BytesOutRate: 18222368
```

```
echo 'show activity' | socat stdio /var/run/haproxy-runtime-api.sock

thread_id: 5 (1..12)
date_now: 1574155030.395933
loops: 1935415309 [ 207397942 157417111 152470256 145664326 181053775 144738936 198399719 146770209 148055392 165906913 143284751 144255979 ]
wake_cache: 1264907055 [ 129956070 103337569 100782300 97093708 115995098 96778384 125374085 97621611 98269053 107734085 95765158 96199934 ]
wake_tasks: 23097052 [ 2091207 1902002 1897785 1875325 1983639 1868456 2044449 1874454 1882239 1932650 1867776 1877070 ]
wake_signal: 0 [ 0 0 0 0 0 0 0 0 0 0 0 0 ]
poll_exp: 0 [ 0 0 0 0 0 0 0 0 0 0 0 0 ]
poll_drop: 111994939 [ 9232919 9307805 9386600 9361072 9342242 9345155 9295655 9334413 9358798 9331125 9334237 9364918 ]
poll_dead: 0 [ 0 0 0 0 0 0 0 0 0 0 0 0 ]
poll_skip: 0 [ 0 0 0 0 0 0 0 0 0 0 0 0 ]
fd_lock: 12771103 [ 1213094 1114770 1039564 949986 1233847 849207 1419890 952495 1003019 1169041 914086 912104 ]
conn_dead: 0 [ 0 0 0 0 0 0 0 0 0 0 0 0 ]
stream: 920678462 [ 76631212 76675385 76937332 76654874 77191848 77121874 77684051 76578258 76210994 76412611 76275688 76304335 ]
pool_fail: 0 [ 0 0 0 0 0 0 0 0 0 0 0 0 ]
buf_wait: 0 [ 0 0 0 0 0 0 0 0 0 0 0 0 ]
empty_rq: 509202025 [ 87420812 38656465 33368717 27183606 61476919 25535139 77494097 28460523 30131249 47355817 25734140 26384541 ]
long_rq: 5690897 [ 466550 478056 473425 470101 485776 469403 477280 477074 471501 474034 478859 468838 ]
ctxsw: 562910429 [ 403975238 404526016 406053603 404620024 407282880 406988137 409502213 404275795 402376357 402915384 402693355 402668723 ]
tasksw: 1175210224 [ 97980073 97849422 98232325 97842863 98519184 98297561 98988952 97698955 97373494 97604661 97391785 97430949 ]
cpust_ms_tot: 60876809 [ 5090697 5072272 5086505 5070698 5108377 4984954 5093190 5078383 5085232 5092787 5073640 5040074 ]
cpust_ms_1s: 16 [ 1 2 0 6 0 0 4 2 0 0 1 0 ]
cpust_ms_15s: 314 [ 48 32 20 20 29 32 29 19 24 7 36 18 ]
avg_loop_us: 66 [ 58 64 59 68 74 61 54 62 67 61 80 84 ]
accepted: 78168525 [ 29659487 3755043 2944132 1995337 8615133 1744248 15829831 2190649 2466372 5303239 1769445 1895609 ]
accq_pushed: 77463382 [ 6458248 6431620 6482065 6462479 6467567 6452476 6455666 6444773 6454638 6445610 6446671 6461569 ]
accq_full: 705143 [ 46442 38674 36666 87552 57108 42905 51059 86046 74553 33328 54508 96302 ]
accq_ring: 0 [ 0 0 0 0 0 0 0 0 0 0 0 0 ]
```

Regarding CPU usage I also tried to fetch some metrics for the time when this started happening, but unfortunately we currently don't have those numbers - right now I can only tell you that the _node_ the pod is running on is fine (CPU usage/load below 40%)

**Comment by joschg** [comment]
I have an update regarding cpu usage: the last few times we saw this happening there actually was a very high load on the node, so seems like the issue actually is correlated to cpu usage.

**Comment by Benjamin-Beeman** [comment]
I am having a similar issue. When I run:
echo 'show info' | socat stdio /var/run/haproxy-runtime-api.sock
or
echo 'show activity' | socat stdio /var/run/haproxy-runtime-api.sock

I get: 2020/01/31 14:52:51 socat[54] E connect(5, AF=1 "/var/run/haproxy-runtime-api.sock", 35): Connection refused

This is also showing up in the logs when the pod starts up:

020/01/31 14:39:05 HAProxy Ingress Controller v1.3.1 e6f6965
2020/01/31 14:39:05 Build from: git@github.com:haproxytech/kubernetes-ingress.git
2020/01/31 14:39:05 Build date: 2020-01-21T20:49:55
2020/01/31 14:39:05 ConfigMap: haproxy-controller/haproxy-configmap
2020/01/31 14:39:05 Ingress class: haproxy
2020/01/31 14:39:05 main.go:97: Default backend service: haproxy-controller/ingress-default-backend
2020/01/31 14:39:05 main.go:98: Default ssl certificate: /
2020/01/31 14:39:05 controller.go:93: Running with  HA-Proxy version 2.0.11 2019/12/11 - https://haproxy.org/
2020/01/31 14:39:05 controller.go:98: Starting HAProxy with /etc/haproxy/haproxy.cfg
2020/01/31 14:39:05 controller.go:111: Running on haproxy-ingress-54fdb7f5f4-w22gv
2020/01/31 14:39:05 configuration.go:93: publish-service: incorrect param
[WARNING] 030/143905 (22) : Can't open server state file '/var/state/haproxy/global': No such file or directory
2020/01/31 14:39:05 controller.go:69: Running on Kubernetes version: v1.14.1 linux/amd64
[ALERT] 030/143905 (22) : Starting frontend https: cannot create listening socket [:::443]
[ALERT] 030/143905 (22) : Starting frontend http: cannot create listening socket [:::80]
2020/01/31 14:39:10 controller.go:175 /var/run/haproxy-runtime-api.sock dial unix /var/run/haproxy-runtime-api.sock: connect: connection refused
2020/01/31 14:39:10 controller-haproxy.go:140: HAProxy reloaded
[WARNING] 030/143910 (36) : Can't open server state file '/var/state/haproxy/global': No such file or directory
[WARNING] 030/143910 (36) : Can't open server state file '/var/state/haproxy/global': No such file or directory
[ALERT] 030/143910 (36) : Starting frontend http: cannot create listening socket [:::80]
[ALERT] 030/143910 (36) : Starting frontend https: cannot create listening socket [:::443]

**Comment by oktalz** [comment]
@Benjamin-Beeman your issue is not related to this, if you read the logs in there you have:
```
cannot create listening socket [:::443]
cannot create listening socket [:::80]
```

**Comment by mshahat** [comment]
is there anyway to configure the chart to deploy and run the haproxy ingress controller with ipv6 disabled?

**Comment by Mo3m3n** [comment]
> is there anyway to configure the chart to deploy and run the haproxy ingress controller with ipv6 disabled?

Available starting from 1.5 via [--disable-ipv6](https://github.com/haproxytech/kubernetes-ingress/blob/master/documentation/controller.md#--disable-ipv6)
